cpp/src/arrow/compute/row/row_internal.cc

```diff
@@ -293,8 +293,12 @@ Status RowTableImpl::ResizeFixedLengthBuffers(int64_t num_extra_rows) {
 }
 
 Status RowTableImpl::ResizeOptionalVaryingLengthBuffer(int64_t num_extra_bytes) {
+  if (metadata_.is_fixed_length) {
+    return Status::OK();
+  }
```

Comment:
IMO it's more elegant to document that `ResizeOptional*VaryingLength*Buffer` has a pre-condition

`/// \pre !metadata_.is_fixed_length`

Then you put a `DCHECK(!metadata_.is_fixed_length)` and review the 3 callsites to ensure this is only called when the pre-condition is respected.

Reply:
In the first callsite it's clearly inside a branch where `!metadata_.is_fixed_length`.

The second callsite should have an if guarding the call to `ResizeOptionalVaryingLengthBuffer` as it doesn't make sense to resize varying length buffers if there aren't any. The client code become more explicit.

```diff
 Status RowTableImpl::AppendEmpty(uint32_t num_rows_to_append,
                                  uint32_t num_extra_bytes_to_append) {
   RETURN_NOT_OK(ResizeFixedLengthBuffers(num_rows_to_append));
-  RETURN_NOT_OK(ResizeOptionalVaryingLengthBuffer(num_extra_bytes_to_append));
+  if (!metadata_.is_fixed_length) {
+    RETURN_NOT_OK(ResizeOptionalVaryingLengthBuffer(num_extra_bytes_to_append));
+  }
   num_rows_ += num_rows_to_append;
   if (metadata_.row_alignment > 1 || metadata_.string_alignment > 1) {
     memset(rows_->mutable_data(), 0, bytes_capacity_);
   }
   return Status::OK();
 }
```

Reply:
Thank you! Addressed.